src/Raven.Server/Documents/Handlers/DocumentHandler.cs

```diff
@@ -108,6 +114,121 @@ public Task Get()
             }
         }
 
+        [RavenAction("/databases/*/bulk_insert", "GET", "/databases/*/bulk_insert")]
+        public async Task BulkInsert()
+        {
+            DocumentsOperationContext context;
+            using (ContextPool.AllocateOperationContext(out context))
+            using (var socket = await HttpContext.WebSockets.AcceptWebSocketAsync())
+            {
+                var list = new List<Document>();
+                try
+                {
+                    var totalSize = 0;
+                    while (true)
+                    {
+                        var task = context.ReadFromWebSocket(socket, "bulk-insert", Server.ServerStore.ServerShutdown);
+                        if (task.IsCompleted == false || totalSize > 16*Voron.Global.Constants.Size.Megabyte)
+                        {
+                            await FlushBatchAsync(list);
+                            totalSize = 0;
+                        }
+
+                        var obj = await task;
+                        if (obj == null)
+                            break;
+                        totalSize += obj.Size;
+
+                        LazyStringValue id;
+                        var data = BuildObjectFromBulkInsertOp(context,obj,out id);
+                        
+                        var doc = new Document
+                        {
+                            Key = id,
+                            Data = data
+                        };
+                        list.Add(doc);
+                    }
+                    await FlushBatchAsync(list);
+                }
+                catch (Exception e)
+                {
+                    using (var ms = new MemoryStream())
+                    using (var writer = new BlittableJsonTextWriter(context, ms))
+                    {
+                        writer.WriteStartObject();
+                        writer.WritePropertyName("Exception");
+                        writer.WriteString(e.ToString());
+                        writer.WriteEndObject();
+                        writer.Flush();
+                        ArraySegment<byte> bytes;
+                        ms.TryGetBuffer(out bytes);
+                        await socket.SendAsync(bytes, WebSocketMessageType.Text, true, Server.ServerStore.ServerShutdown);
+                    }
+                }
+                finally
+                {
+                    await socket.CloseAsync(WebSocketCloseStatus.NormalClosure, "Bulk-insert", Server.ServerStore.ServerShutdown);
+                }
+            }
+        }
+
+        private static BlittableJsonReaderObject BuildObjectFromBulkInsertOp(JsonOperationContext ctx, BlittableJsonReaderObject obj,out LazyStringValue id)
```

Comment:
Bad idea, since this requires us to build the document several times.
Instead, see how we use BatchRequestParser to only parse things once.